lib/langchain/tool/database.rb

```diff
@@ -43,39 +47,66 @@ def initialize(connection_string:, tables: [], exclude_tables: [])
     end
 
     # Database Tool: Returns a list of tables in the database
+    #
+    # @return [Array<Symbol>] List of tables in the database
     def list_tables
       db.tables
     end
 
     # Database Tool: Returns the schema for a list of tables
     #
-    # @param tables [String] The tables to describe.
-    # @return [String] Database schema for the tables
-    def describe_tables(tables:)
-      schema = ""
-      tables.split(",").each do |table|
-        describe_table(table, schema)
-      end
-      schema
+    # @param tables [Array<String>] The tables to describe.
+    # @return [String] The schema for the tables
+    def describe_tables(tables: [])
+      return "No tables specified" if tables.empty?
+
+      Langchain.logger.info("Describing tables: #{tables}", for: self.class)
+
+      tables
+        .map do |table|
+          describe_table(table)
+        end
+        .join("\n")
     end
 
     # Database Tool: Returns the database schema
     #
     # @return [String] Database schema
     def dump_schema
       Langchain.logger.info("Dumping schema tables and keys", for: self.class)
-      schema = ""
-      db.tables.each do |table|
-        describe_table(table, schema)
+
+      schemas = db.tables.map do |table|
+        describe_table(table)
       end
-      schema
+      schemas.join("\n")
     end
 
-    def describe_table(table, schema)
+    # Database Tool: Executes a SQL query and returns the results
+    #
+    # @param input [String] SQL query to be executed
+    # @return [Array] Results from the SQL query
+    def execute(input:)
+      Langchain.logger.info("Executing \"#{input}\"", for: self.class)
+
+      db[input].to_a
+    rescue Sequel::DatabaseError => e
+      Langchain.logger.error(e.message, for: self.class)
+      e.message # Return error to LLM
```

Comment:
This is crucial because we return back the error to the LLM to attempt to self-correct.